src/orderbook/__tests__/state.spec.ts

```diff
@@ -0,0 +1,124 @@
+import { EventData } from 'web3-eth-contract'
+import { DEFAULT_ORDERBOOK_OPTIONS } from '..'
+import { AccountState } from '../state'
+
+function accountState(): AccountState {
+  return new AccountState(
+    { forEvent: (ev) => Promise.resolve(ev.blockNumber) },
+    { ...DEFAULT_ORDERBOOK_OPTIONS, strict: true },
+  )
+}
+
+function ev(batch: number, name: string, data: any, index: number = 0): EventData {
+  return {
+    ...{} as EventData,
+    event: name,
+    blockNumber: batch,
+    returnValues: data,
+    logIndex: index || 0,
+  }
+}
+
+function addr(lowerBits: number): string {
+  const hex = lowerBits.toString(16)
+  return `0x${hex.padStart(40 - hex.length, '0')}`
+}
+
+describe('Account State', () => {
+  describe('applyEventsUntilBatch', () => {
+    it('Updates the batch ID', async () => {
+      const state = accountState()
+      await state.applyEventsUntilBatch(
+        [ev(1, 'TokenListing', { id: '0', token: addr(1) })],
+        1,
+      )
+      expect(state.toJSON().batchId).toBe(1)
+    })
+
+    it('Throws when events are out of order', async () => {
+      const state = accountState()
+      expect(state.applyEventsUntilBatch(
+        [
+          ev(2, 'TokenListing', { id: '0', token: addr(1) }),
+          ev(1, 'TokenListing', { id: '0', token: addr(1) }),
+        ],
+        2,
+      )).rejects.toThrow()
+    })
+
+    it('Throws when events are outside expected batch range', async () => {
+      const state = accountState()
+      state.applyEventsUntilBatch([], 1)
+      expect(state.applyEventsUntilBatch(
+        [ev(1, 'TokenListing', { id: '0', token: addr(1) })],
+        2,
+      )).rejects.toThrow()
+      expect(state.applyEventsUntilBatch(
+        [ev(3, 'TokenListing', { id: '0', token: addr(1) })],
+        2,
+      )).rejects.toThrow()
+    })
+  })
+
+  describe('OrderPlacement > OrderCancellation > OrderDeletion', () => {
+    it('Sets the order to become invalid in the current batch', async () => {
```

Comment:
Nit: no strong feelings, but this could also be three test cases as unit tests should usually only test one thing.